### DISCUSSION TITLE
Mock with a relative path does not work

### DISCUSSION THREAD

**Discussion by chawax** [vitest-dev/vitest#4927] [Q&A]
I try to write a mock a hook (exported as default from a file) this way : 

```
    vi.mock('../../hooks/auth/useUser', () => {
      return {
        default: () => ({
          isUser: false,
          isManager: true,
        }),
      };
    });
```

But the mock does not work

If I move the hook file to the same directory and write the mock this way :

```
    vi.mock('./useUser', () => {
      return {
        default: () => ({
          isUser: false,
          isManager: true,
        }),
      };
    });
```

It works.

So the problem is mocking a hook using a relative path but I am sure about the path.

Tried with latest Vitest (1.1.3) and Vite (5.0.11)

Anyone had a similar problem ?

**Comment by code-by-gijs** [comment]
Yes! same issue here (sort of)

this doesnt work:
```
vitest.mock('~shared/ui/hooks/use-analytics');
```

but this works:
```
vitest.mock('~shared/ui/hooks/use-analytics', () => {
    return {
        default: () => ({
            event: () => true,
            pageView: () => true,
            initialEvent: () => true,
            gaId: '123',
        }),
    };
});
```

`~shared` is an alias in my vite.config.mts:
```
    resolve: {
        alias: {
            '~shared': path.resolve(__dirname, '../../libs/shared/source'),
        },
    },
```

**Reply by hi-ogawa** [reply]
@code-by-gijs Can you elaborate what you mean by "this doesn't work"?

If you use `vitest.mock` without a factory function `() => ...`, then what Vitest does is to make a simple fake exports like `{ default: () => undefined }` as explained as "automocking" https://vitest.dev/guide/mocking.html#automocking-algorithm

So, if you want to customize function return value, then it's expected that you need to provide a factory function like you do in "this works".

**Reply by code-by-gijs** [reply]
I have a mocked use-analytics.tsx. It's in the a __mocks__ folder. It throws an error saying the mock file can't be found at a certain path, but the file exists. 

Meanwhile, it's fixed. Not sure what fixed it but it seems to have something to do with relative paths.

**Reply by hi-ogawa** [reply]
> I have a mocked use-analytics.tsx. It's in the a mocks folder. It throws an error saying the mock file can't be found at a certain path, but the file exists.

Thanks for providing the detail. Yeah, that might be a bug as well (maybe related to alias?), but cannot tell easily without a reproduction.
Maybe worth investigating, but glad to know you got it fixed.

**Reply by code-by-gijs** [reply]
Sorry, it's still failing when using it like this:
```
vitest.mock('~shared/ui/hooks/use-analytics');

Error: Failed to load url /home/vsts/work/1/s/libs/shared/source/ui/hooks/__mocks__/use-analytics.tsx (resolved id: /home/vsts/work/1/s/libs/shared/source/ui/hooks/__mocks__/use-analytics.tsx). Does the file exist?
```
the file exists! I have multiple mocks like this. It used to work when using an older version of vitest (I upgraded from 0.34.6 to latest)

**Reply by code-by-gijs** [reply]
So this is what does work:

```
import analyticsMock from '~shared/ui/hooks/__mocks__/use-analytics';

vitest.mock('~shared/ui/hooks/use-analytics', () => {
    return {
        default: () => analyticsMock(),
    };
});

**Reply by hi-ogawa** [reply]
I couldn't reproduce the issue with testing things like:

```tsx
export default defineConfig({
  test: {
    alias: {
      "~src": __dirname + "/src",
      "~deep": path.resolve(__dirname, "../deep"),
    }
  },
})
```

We would need a reproduction to investigate further, but your workaround might be fine.
I think you might be able to shorten it further by dynamic import:

```ts
vitest.mock('~shared/ui/hooks/use-analytics', () =>
  import('~shared/ui/hooks/__mocks__/use-analytics')
);
```

**Comment by hi-ogawa** [comment]
> So the problem is mocking a hook using a relative path but I am sure about the path.

@chawax if this is only the difference, then I think that could be a bug. Other possible cause is that somehow import order differs between two cases and `useUser` module is imported before `vi.mock` kicks in https://vitest.dev/guide/common-errors.html#cannot-mock-mocked-file-js-because-it-is-already-loaded (though this error is already reverted https://github.com/vitest-dev/vitest/pull/5050, it can happen that mock is silently not working by the same reason).

We have many test cases with `vi.mock("../src/something")` in https://github.com/vitest-dev/vitest/tree/main/examples/mocks/test, so it's expected work. It would be nice if you can come with a reproduction to show in which case mock is failing.
